client_server/tcp_full_duplex_client.c

```diff
@@ -0,0 +1,174 @@
+/**
+ * @file
+ * @author [NVombat](https://github.com/NVombat)
+ * @brief Client-side implementation of [TCP Full Duplex
+ * Communication](http://www.tcpipguide.com/free/t_SimplexFullDuplexandHalfDuplexOperation.htm)
+ * @see tcp_full_duplex_server.c
+ *
+ * @details
+ * The algorithm is based on the simple TCP client and server model. However,
+ * instead of the server only sending and the client only receiving data,
+ * The server and client can both send and receive data simultaneously. This is
+ * implemented by using the `fork` function call so that in the server the child
+ * process can recieve data and  parent process can send data, and in the client
+ * the child process can send data and the parent process can receive data. It
+ * runs an infinite loop and can send and receive messages indefinitely until
+ * the user exits the loop. In this way, the Full Duplex Form of communication
+ * can be represented using the TCP server-client model & socket programming
+ */
+
+#include <arpa/inet.h>             /// For the type in_addr_t and in_port_t
+#include <netdb.h>                /// For structures returned by the network database library - formatted internet addresses and port numbers
+#include <netinet/in.h>          /// For in_addr and sockaddr_in structures
+#include <stdint.h>             /// For specific bit size values of variables
+#include <stdio.h>             /// Variable types, several macros, and various functions for performing input and output
+#include <stdlib.h>           /// Variable types, several macros, and various functions for performing general functions
+#include <string.h>          /// Various functions for manipulating arrays of characters
+#include <sys/socket.h>     /// For macro definitions related to the creation of sockets
+#include <sys/types.h>     /// For definitions to allow for the porting of BSD programs
+#include <unistd.h>       /// For miscellaneous symbolic constants and types, and miscellaneous functions
+
+
+#define PORT 10000  /// Define port over which communication will take place
+
+/**
+ * @brief Utility function used to print an error message to `stderr`.
+ * It prints `str` and an implementation-defined error
+ * message corresponding to the global variable `errno`.
+ * @returns void
+ */
+void error()
+{
+    perror("Socket Creation Failed");
+    exit(EXIT_FAILURE);
+}
+
```

Comment:
_If applicable_: add self-test implementations using `assert`.

Reply:
It wont be applicable in this case as the only user input/variable data is the message being sent which can be any format or string anyway